numba/np/npdatetime.py

```diff
@@ -812,3 +814,29 @@ def _cast_to_timedelta(context, builder, val):
 @lower_builtin(np.isnat, types.NPTimedelta)
 def _np_isnat_impl(context, builder, sig, args):
     return npyfuncs.np_datetime_isnat_impl(context, builder, sig, args)
+
+
+@lower_cast(types.NPDatetime, types.Integer)
+@lower_cast(types.NPTimedelta, types.Integer)
+def _cast_npdatetime_int64(context, builder, fromty, toty, val):
+    if toty.bitwidth != 64: # all date time types are 64 bit
+        msg = f"Cannot cast {fromty} to {toty} as {toty} is not 64 bits wide."
+        raise ValueError(msg)
+    return val
+
+
+@overload_method(types.NPTimedelta, '__hash__')
+@overload_method(types.NPDatetime, '__hash__')
+def ol_hash_npdatetime(inst):
+    if IS_32BITS:
+        def impl(inst):
+            hi = (np.int64(inst) & 0xffffffff00000000)
```

Comment:
```suggestion
            hi = (np.int64(inst) & 0xffffffff00000000) >> 32
```

Reply:
If I shift `hi` to the right, then, `f(TD('-4'))` fails

Reply:
I think the types will need to be got exactly right, this might be quicker to do using `builder` or maybe `@njit(locals={})`.

Reply:
Hum... is there any particular reason for using `locals={}`?

Reply:
Apologies, `{}` is indicative only, you'd need to populate it appropriately. It gives control of the type of variables. There's some examples in https://github.com/numba/numba/blob/master/numba/cpython/hashing.py#L127-L160

Reply:
Good to know! By the way, the PR is ready for review.

Reply:
Thanks, I'll move the label.